runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PipelineResourcesTest.java

```diff
@@ -44,25 +46,68 @@
   @Rule public transient ExpectedException thrown = ExpectedException.none();
 
+  ClassLoader currentContext = null;
+
+  @Before
+  public void setUp() {
+    currentContext = Thread.currentThread().getContextClassLoader();
+  }
+
+  @After
+  public void tearDown() {
+    Thread.currentThread().setContextClassLoader(currentContext);
+  }
+
   @Test
-  public void detectClassPathResourceWithFileResources() throws Exception {
+  public void detectClassPathResourceWithFileResources() throws IOException {
     File file = tmpFolder.newFile("file");
     File file2 = tmpFolder.newFile("file2");
     URLClassLoader classLoader =
-        new URLClassLoader(new URL[] {file.toURI().toURL(), file2.toURI().toURL()});
+        new URLClassLoader(
+            new URL[] {file.toURI().toURL(), file2.toURI().toURL()},
+            Thread.currentThread().getContextClassLoader());
+
+    Thread.currentThread().setContextClassLoader(classLoader);
+
+    List<String> detected = PipelineResources.detectClassPathResourcesToStage(getClass());
+    assertTrue(detected.contains(file.getAbsolutePath()));
+    assertTrue(detected.contains(file2.getAbsolutePath()));
+
+    // locate java home and verify that we didn't pull any resource from JDK
+    String javaHome =
+        Optional.ofNullable(System.getProperty("java.home"))
+            .map(File::new)
+            .map(File::getAbsolutePath)
+            .orElse(null);
+    assertFalse(detected.stream().anyMatch(f -> f.startsWith(javaHome)));
+  }
+
+  @Test
+  public void detectClassPathResourcesFromHierarchy() throws IOException {
+    File file = tmpFolder.newFile("file");
+    File file2 = tmpFolder.newFile("file2");
+    URLClassLoader classLoader1 =
+        new URLClassLoader(
+            new URL[] {file.toURI().toURL()}, Thread.currentThread().getContextClassLoader());
+    ClassLoader classLoader2 = new ClassLoader(classLoader1) {};
+
+    URLClassLoader classLoader3 =
+        new URLClassLoader(new URL[] {file2.toURI().toURL()}, classLoader2);
 
-    assertEquals(
-        ImmutableList.of(file.getAbsolutePath(), file2.getAbsolutePath()),
-        PipelineResources.detectClassPathResourcesToStage(classLoader));
+    Thread.currentThread().setContextClassLoader(classLoader3);
+
+    List<String> detected = PipelineResources.detectClassPathResourcesToStage(getClass());
+    assertTrue(detected.contains(file.getAbsolutePath()));
+    assertTrue(detected.contains(file2.getAbsolutePath()));
   }
 
   @Test
   public void detectClassPathResourcesWithUnsupportedClassLoader() {
     ClassLoader mockClassLoader = Mockito.mock(ClassLoader.class);
     thrown.expect(IllegalArgumentException.class);
     thrown.expectMessage("Unable to use ClassLoader to detect classpath elements.");
-
-    PipelineResources.detectClassPathResourcesToStage(mockClassLoader);
+    Thread.currentThread().setContextClassLoader(mockClassLoader);
+    PipelineResources.detectClassPathResourcesToStage(getClass());
```

Comment:
Not sure why we throw here. According to the current logic, if there was one other UrlClassloader, we would not fail.

Reply:
Same reasoning as above.